pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,681 @@
+	}
+	oldGroupMemberSet, exist := c.setByGroup[group.Name]
+	if !exist {
+		return nil
+	}
+	removedMembers := make([]v1beta1.GroupMember, 0, len(group.GroupMembers))
+	NewGroupMemberPodSet := v1beta1.GroupMemberSet{}
+
+	for i := range groupMemberSet {
+		if oldGroupMemberSet.Has(groupMemberSet[i]) {
+			removedMembers = append(removedMembers, *groupMemberSet[i])
+		} else {
+			NewGroupMemberPodSet.Insert(groupMemberSet[i])
+		}
+	}
+	c.setByGroup[group.Name] = NewGroupMemberPodSet
+	c.onEgressRemove(removedMembers)
+	return nil
+}
+
+func (c *Controller) onEgresssAdd(groupName string, members []v1beta1.GroupMember) {
+	ip, exist := c.policyMap[groupName]
+	if !exist {
+		return
+	}
+	for i := range members {
+		c.enqueueRuleAdd(ip, &members[i])
+	}
+}
+
+func (c *Controller) onEgressRemove(members []v1beta1.GroupMember) {
+	for i := range members {
+		c.enqueueRuleDelete(&members[i])
+	}
+}
+
+func (c *Controller) onEgressGroupAdd(groupName string) {
+	ip := c.policyMap[groupName]
+	groups, _ := c.setByGroup[groupName]
+	for _, member := range groups {
+		c.enqueueRuleAdd(ip, member)
+	}
+}
+
+func (c *Controller) onEgressGroupRemove(groupName string) {
+	groups, _ := c.setByGroup[groupName]
+	for _, member := range groups {
+		c.enqueueRuleDelete(member)
+	}
+}
+
+func (c *Controller) addEgressPolicy(obj interface{}) {
+	c.policyMapLock.Lock()
+	defer c.policyMapLock.Unlock()
+
+	egress := obj.(*v1alpha1.Egress)
+	egressIP := egress.Spec.EgressIP
+	klog.Infof(egressIP)
+	if isLocalIP(egressIP) {
+		c.IPAllocatorLock.Lock()
+		defer c.IPAllocatorLock.Unlock()
+		id, err := c.IPAllocator.allocateForIP(egressIP)
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		c.ofClient.InstallSNATMarkFlows(net.ParseIP(egressIP), id)
+		if err != nil {
+			klog.Errorf(err.Error())
+			return
+		}
+		errroute := c.routeClient.AddSNATRule(net.ParseIP(egressIP), id)
+		if errroute != nil {
+			klog.Infof(errroute.Error())
+			return
+		}
+	}
+	c.addEgressPolicyLocked(egress)
+}
+
+func (c *Controller) updateEgressPolicy(old, cur interface{}) {
+	c.policyMapLock.Lock()
+	defer c.policyMapLock.Unlock()
+	egressOld := old.(*v1alpha1.Egress)
+	egressNew := cur.(*v1alpha1.Egress)
+
+	newEgressIP := egressNew.Spec.EgressIP
+	oldEgressIP := egressOld.Spec.EgressIP
+	if newEgressIP == oldEgressIP {
+		return
+	}
+	if isLocalIP(oldEgressIP) {
+		c.IPAllocatorLock.Lock()
+		defer c.IPAllocatorLock.Unlock()
+		id, errRelease := c.IPAllocator.release(oldEgressIP)
+		klog.Infof("%#v\n", c.IPAllocator)
+
+		if errRelease != nil {
+			klog.Errorf(errRelease.Error())
+			return
+		}
+		c.ofClient.UninstallSNATMarkFlows(id)
+		errroute := c.routeClient.DeleteSNATRule(id)
+		if errroute != nil {
+			klog.Infof(errroute.Error())
+			return
+		}
+
+	}
+	if isLocalIP(newEgressIP) {
+		id, errAllocate := c.IPAllocator.allocateForIP(newEgressIP)
+		klog.Infof("%#v\n", c.IPAllocator)
+
+		if errAllocate != nil {
+			c.IPAllocator.allocateForIP(oldEgressIP)
```

Comment:
Is this kind of retry? Doesn't seem elegant to me. And many error handling is missing here as well as other places in this file.